### PR TITLE
Stamping

### LINKED ISSUE #29: Support stamp variables
Not sure what facilities exist at the `jsonnet` level for injecting this kind of thing, but especially in the context of Kubernetes templating it would be fantastic if we could optionally inject build context.

The particular scenario I am envisioning would be environment modeling, where I might want to inject `{BUILD_USER}` into the template expansion (e.g. as a Docker Image tag).  Or for integration testing on PRs I might want an environment per PR based on the PR number, which might come through `--workspace_status_command`.

### LINKED ISSUE #49: Unable to use ext_strs with literal key-value pairs
Based on the documentation, I expect to be able to pass in literal key-value pairs in the `ext_strs` dict:

> Map of strings to pass to jsonnet as external variables via `--ext-str key=value`.

and not names of configuration variables.

However, at [jsonnet.bzl, line 118](https://github.com/bazelbuild/rules_jsonnet/blob/6dd47d83adda2dc7fce24a7b45dee202d7575625/jsonnet/jsonnet.bzl#L118), the value of the external string is taken using `_quote(ctx.var[val])`. Is there a reason `_quote(val)` was not used instead?

### PR DESCRIPTION
**Pull request by Globegitter**
Fixes #29 - at least for external strings, it does not support stamping for external code, but the current implementation should support the use-cases mentioned in the issue and should also support our own use-cases. I also noticed that https://github.com/bazelbuild/rules_jsonnet/commit/6dd47d83adda2dc7fce24a7b45dee202d7575625 broke the make variable support - I fixed it and added a test for it. So should also fix #49

The stamper.py is taken directly from: https://github.com/bazelbuild/rules_k8s/blob/master/k8s/stamper.py

Edit: I have now added support for stamping for `ext_code` and stamping has to be explicitly activated, similar to how `rules_docker` does it by setting `stamp = True`. 

Edit2: There where still some issues with the approach I took before (`stamp = True`), since the stamper is replacing variables within `{}`, so if you also where passing in JSON that would cause issues, so I now changed it that the keys that should be stamped have to be explicitly named, via `stamp_keys=['key']`

### CODE REVIEW

**Comment by borg286** [comment]
I am not the owner of jsonnet bazel library, but am interested in getting this PR integrated. Do you think you could resolve these conflicts so as to be prepared for when the owner does get around to accepting your PR?

File: jsonnet/jsonnet.bzl [hunk 1]
```diff
@@ -115,12 +160,12 @@ def _jsonnet_to_json_impl(ctx):
        "-J %s" % ctx.genfiles_dir.path,
        "-J %s" % ctx.bin_dir.path] +
       ["--ext-str %s=%s"
-       % (_quote(key), _quote(ctx.var[val])) for key, val in jsonnet_ext_strs.items()] +
+       % (key, _quote(val)) for key, val in jsonnet_ext_strs.items()] +
```
Comment (borg286) [inline jsonnet/jsonnet.bzl]: Why is it safe to not quote the key?

Comment (Globegitter) [reply jsonnet/jsonnet.bzl]: Good question - I will take a look at that again.

Comment (Globegitter) [reply jsonnet/jsonnet.bzl]: You are right, we should quote keys, in case they contain quotes. I added a test case for this.

**Comment by Globegitter** [comment]
@borg286 I updated to latest master and updated the API of it - we are using this feature ourselves now internally but let me know if you run into any issues.

File: examples/BUILD [hunk 2]
```diff
@@ -110,6 +124,39 @@ genrule(
     cmd = "echo {} > $@"
 )
 
+jsonnet_to_json(
```
Comment (bshi) [inline examples/BUILD]: This rule is redundant with the accompanying jsonnet_to_json_test rule.

File: examples/BUILD [hunk 3]
```diff
@@ -88,6 +95,13 @@ jsonnet_to_json_test(
     name = "files_test",
     size = "small",
     src = "files.jsonnet",
+    ext_strs = {
```
Comment (bshi) [inline examples/BUILD]: These changes appear unrelated to stamping?

File: examples/BUILD (hunk 2 above)
Comment (Globegitter) [reply examples/BUILD]: fixed now based on changes in latest master.

File: examples/BUILD (hunk 3 above)
Comment (Globegitter) [reply examples/BUILD]: removed unrelated tests.

File: jsonnet/jsonnet.bzl [hunk 4]
```diff
@@ -307,7 +356,9 @@ def _jsonnet_to_json_test_impl(ctx):
       list(transitive_data) +
       list(depinfo.transitive_sources) +
       [list(jfile.files)[0] for jfile in jsonnet_ext_str_files] +
-      [list(jfile.files)[0] for jfile in jsonnet_ext_code_files])
+      [list(jfile.files)[0] for jfile in jsonnet_ext_code_files] +
```
Comment (laurentlb) [inline jsonnet/jsonnet.bzl]: Feel free to ignore the comment (issue was already present), but this looks suspicious: `list(jfile.files)[0]`
(is it possible to have multiple files in the list?)

Comment (Globegitter) [reply jsonnet/jsonnet.bzl]: Yeah that was my code and when I wrote this originally I was not aware of idiomatic usage. This PR https://github.com/bazelbuild/rules_jsonnet/pull/67 should fix that though. If you also have the time to look over that one that would be very much appreciated.

**Comment by laurentlb** [comment]
Formatting of the README.md looks broken (https://github.com/Globegitter/rules_jsonnet/blob/873f9845c33e85c746f738a094a55f92559fa6ac/README.md).

**Comment by mtsgrd** [comment]
@laurentlb I created https://github.com/bazelbuild/rules_jsonnet/pull/74 for adding a missing `</pre>` tag.
